lib/fog/core/cache.rb

```diff
@@ -0,0 +1,183 @@
+  #
+  # Find references of yaml in this file, split out to strategy objects/diff backends etc.
+  class Cache
+    class CacheNotFound < StandardError; end
+
+    # when a resource is used such as `server.cache.dump` the model klass is passed in
+    # so that it can be identified from a different session.
+    attr_reader :model
+
+    # Loads cache associated to the +model_klass+ and +service+ into memory.
+    #
+    # If no cache is found, it will raise an error for handling:
+    #
+    #   rescue Fog::Cache::CacheNotFound
+    #     set_initial_cache
+    #
+    def self.load(model_klass, service)
+      cache_files = Dir.glob("#{sandbox(model_klass, service)}/*")
+
+      raise CacheNotFound if cache_files.empty?
+
+      loaded = cache_files.map do |path|
+        model_klass = Object.const_get(cache_data(path)[:model_klass])
+        model_klass.new(cache_data(path)[:attrs])
+      end
+
+      collection_klass = cache_data(cache_files.sample)[:collection_klass] &&
+                         Object.const_get(cache_data(cache_files.sample)[:collection_klass])
+
+      loaded.each do |i|
+        # See https://github.com/fog/fog-aws/issues/354#issuecomment-286789702
+        i.collection = collection_klass.new(:service => service) if collection_klass
+        i.instance_variable_set(:@service, service)
+      end
+    end
+
+    # Expires cache - this does not expire all cache associated.
+    # Instead, this will remove all on-disk cache of this specific +model_klass+ and and +@service+
+    def self.expire_cache!(model_klass, service)
+      FileUtils.rm_rf(sandbox(model_klass, service))
+    end
+
+    # loads yml cache from path on disk
+    def self.cache_data(path)
+      @cache_data ||= {}
+      return @cache_data[path] if @cache_data[path]
+      @cache_data[path] = YAML.load(File.read(path))
+    end
+
+    # The path/namespace where the cache is stored for a specific +model_klass+ and +@service+.
+    def self.sandbox(model_klass, service)
+      service_klass = service.class
+      name = File.join(Dir.tmpdir, service_klass.to_s, service_identifier(service), model_klass.to_s)
+      name = safe_path(name)
+      FileUtils.mkdir_p(name)
+      name
+    end
+
+    # Used for the cache namespace as a way to
+    # segregate caches used from different cloud accounts
+    # and different resources/services within those.
+    #
+    # ideally fog-core services had an identifier similar to
+    # `identity` for model classes, but until then this will do.
+    def self.service_identifier(service)
+
+      region = service.region # in a specific region
+      klass = safe_path(service.class) # for this set of resources (elbs/compute/asgs etc)
+
+      key_id = if service.instance_variable_get(:@aws_access_key_id)
+                 # for a specific api key: constrain to a specific amazon account/user.
+                 service.instance_variable_get(:@aws_access_key_id)
```

Comment:
I thought this over last night a bit more. This need basically comes out of needing to create a unique sandbox for a given setup: ie, for a given user, a given service account (google/aws/whatever), region, etc we need to make sure the cache doesn't step over other setups. Instead of having this auto detection, it may be more desirable to push this down to a configurable. Who knows what namespacing constraints would be needed. It's possible using STS tokens or Api keys this uniqeness constraint would be determined otherwise and not just api tokens. Users could then define this constraint in their code via an api like:

```
Fog::Cache.namespace = "I-determine-my-sandbox-path"
```
 
Then the cache data associated would go to `~/.fog/I-determine-my-sandbox-path/` . This simplifies this cache class quite a bit and would actually be more robust. I'll push a commit soon.